### PR TITLE
Added the logic for Handler#load

### PR DESCRIPTION
**Pull request by SebastiaanYN**
ESLint now errors when a function/method is async but doesn't use await.

Handler#load now contains the functionality for loading command/event files from a directory. It does this recursively to improve possible file structures. It does a lot of checks to make sure the files are correct, and provides useful error messages when something is wrong.

### CODE REVIEW

File: src/handler/Command.js [hunk 1]
```diff
@@ -24,6 +24,9 @@ class Command {
       throw new TypeError('Usage must be a string');
     }
     this.usage = options.usage;
+
+    // Category will be assigned by the Handler
+    this.category = null;
```
Comment (SebastiaanYN) [inline src/handler/Command.js]: I personally think this could use some improvements in the future.

File: .eslintrc.json [hunk 2]
```diff
@@ -6,6 +6,7 @@
   "rules": {
     "no-console": "off",
     "class-methods-use-this": "off",
+    "require-await": "error",
```
Comment (DarkSeraphim) [inline .eslintrc.json]: Please move this to it's own PR. I'd like to keep the PRs centered around a single change if possible, for history purposes.

(i.e. if we need to revert this change, we don't want to accidentally disable the check)

File: src/Utils.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
+const fs = require('fs');
+const path = require('path');
+
+class Utils {
+  /**
+   * @description Read a directory recursively to get all files
+   * @param {string} directory The directory to read
+   * @returns {Array<string>} All the paths to the files
+   */
+  static readdirSyncRecursive(directory) {
```
Comment (DarkSeraphim) [inline src/Utils.js]: Why sync and not async?

File: src/handler/Handler.js [hunk 4]
```diff
@@ -1,26 +1,115 @@
...
+        .forEach((file) => {
+          // eslint-disable-next-line global-require, import/no-dynamic-require
+          const CommandOrEvent = require(file);
+
+          // Try to instantiate the command or event, throw an error if it fails
+          let commandOrEvent;
+          try {
+            commandOrEvent = new CommandOrEvent(dependencies);
+          } catch (err) {
+            throw new Error(`Module '${file}' could not be instantiated, does it export a class?`);
```
Comment (DarkSeraphim) [inline src/handler/Handler.js]: This would prevent someone from writing their own utilities in separate files. Is that wanted behaviour?

File: src/handler/Handler.js [hunk 5]
```diff
@@ -1,26 +1,115 @@
...
+
+          // Handle the object, throw an error if it does not extend Command or Event
+          if (commandOrEvent instanceof Command) {
+            commandOrEvent.category = category;
+
+            this.loadCommand(commandOrEvent, file);
+          } else if (commandOrEvent instanceof Event) {
+            this.loadEvent(commandOrEvent);
+          } else {
+            throw new Error(`Module '${file}' does not extend Command or Event`);
```
Comment (DarkSeraphim) [inline src/handler/Handler.js]: This would prevent someone from writing their own utilities in separate files. Is that wanted behaviour?

File: src/handler/Handler.js [hunk 6]
```diff
@@ -1,26 +1,115 @@
...
+   */
+  loadCommand(command, file) {
+    // Command name might be in use or name might already be an existing alias
+    if (this.commands.has(command.name) || this.aliases.has(command.name)) {
+      throw new Error(`Can't load '${file}', the name '${command.name}' is already used as a command name or alias`);
+    }
+
+    this.commands.set(command.name, command);
+
+    command.aliases.forEach((alias) => {
```
Comment (DarkSeraphim) [inline src/handler/Handler.js]: I'm a bit late to the party, but do we want to force aliases to be set? If not, I'd personally recommend setting it to an empty array so commands that only have a name and no aliases don't have to do anything with aliases.

File: src/handler/Handler.js [hunk 7]
```diff
@@ -1,26 +1,115 @@
...
+
+      this.aliases.set(alias, command);
+    });
+  }
+
+  /**
+   * @description Load an event
+   * @param {Event} event The event that needs to be loaded
+   */
+  loadEvent(event) {
```
Comment (DarkSeraphim) [inline src/handler/Handler.js]: Suggestion: check if the event exists in discord.js 😛. That way you can throw exceptions instead of wondering why your events don't work. (There's an object of them in Discord.Constants iirc)

File: .eslintrc.json (hunk 2 above)
Comment (SebastiaanYN) [reply .eslintrc.json]: Moved to #22

File: src/Utils.js (hunk 3 above)
Comment (SebastiaanYN) [reply src/Utils.js]: This function is called from `Handler#load`. `Handler#load` will often, if not always, be called from the top level of a file, which doesn't have async be default. In this case blocking isn't an issue, since there isn't much else that would have to run during this call.

File: src/handler/Handler.js (hunk 4 above)
Comment (SebastiaanYN) [reply src/handler/Handler.js]: This is intended behaviour, since it makes it clear that something is probably not setup correctly. This could be changed to just logging to console, but that would be less strict.

A stricter option would be having a way of ignoring specific files/folder, since that would allow for utils to be inside the modules folder.

File: src/handler/Handler.js (hunk 6 above)
Comment (SebastiaanYN) [reply src/handler/Handler.js]: Currently you would have to set the aliases to an empty array manually, this can be changed to just defaulting to an empty array when not set.

File: src/handler/Handler.js (hunk 4 above)
Comment (DarkSeraphim) [reply src/handler/Handler.js]: I honestly don't like the fact that people can't write utility classes. I'd opt to just ignore the files without logging, as it's not necessarily an error. People will notice when their module doesn't load.

Alternatively, why not write a test for it?

File: src/handler/Handler.js (hunk 6 above)
Comment (DarkSeraphim) [reply src/handler/Handler.js]: Given that we now throw an error if it's not an array, it is probably fine for this PR to stay like this and to resolve this in a separate PR.

File: src/handler/Command.js (hunk 1 above)
Comment (DarkSeraphim) [reply src/handler/Command.js]: Why not set it manually? Or provide a require/static field with a predetermined set of categories?

Then just throw an exception of it's not a valid option.

File: src/handler/Handler.js (hunk 7 above)
Comment (DarkSeraphim) [reply src/handler/Handler.js]: Just for resolving sake, do you intend to add this in this PR? Or do I mark it solved and leave it for another?

Comment (SebastiaanYN) [reply src/handler/Handler.js]: This is something for a future PR

File: src/handler/Command.js (hunk 1 above)
Comment (SebastiaanYN) [reply src/handler/Command.js]: Determing the category from the folder name would make it more error proof. It also makes everything inside the folder the same category, which is how this structure should work.

File: src/handler/Handler.js (hunk 4 above)
Comment (SebastiaanYN) [reply src/handler/Handler.js]: > Alternatively, why not write a test for it?
What do you mean by this?

Comment (DarkSeraphim) [reply src/handler/Handler.js]: instead of forcing all files to be commands or events, ignoring them if they're not. Then, in a test, you check if all modules that you expect to be loaded, are loaded.

Comment (SebastiaanYN) [reply src/handler/Handler.js]: I believe that makes it a bit too complicated. I prefer just ignoring the files in that case.

Comment (DarkSeraphim) [reply src/handler/Handler.js]: Could you remove the error throwing here and the throw further down in that case? In case you still want logging, I'd like to suggest debug logging.
